src/psyclone/parse/algorithm.py

```diff
@@ -583,29 +660,68 @@ def get_kernel(parse_tree, alg_filename):
             # A simple variable e.g. arg
             full_text = str(argument).lower()
             var_name = full_text
-            arguments.append(Arg('variable', full_text, var_name))
+            try:
+                datatype = arg_type_defns[var_name]
+            except KeyError:
+                datatype = None
+            arguments.append(Arg('variable', full_text, varname=var_name,
+                                 datatype=datatype))
         elif isinstance(argument, Part_Ref):
             # An indexed variable e.g. arg(n)
             full_text = argument.tostr().lower()
             var_name = str(argument.items[0]).lower()
-            arguments.append(Arg('indexed_variable', full_text, var_name))
+            try:
+                datatype = arg_type_defns[var_name]
+            except KeyError:
+                datatype = None
+            arguments.append(Arg('indexed_variable', full_text,
+                                 varname=var_name, datatype=datatype))
         elif isinstance(argument, Function_Reference):
-            # A function reference e.g. func()
+            # A function reference e.g. func(). The datatype of this
+            # function is not determined so datatype in Arg is None.
             full_text = argument.tostr().lower()
             designator = argument.items[0]
             lhs = designator.items[0]
             lhs = create_var_name(lhs)
             rhs = str(designator.items[2])
             var_name = "{0}_{1}".format(lhs, rhs)
             var_name = var_name.lower()
-            arguments.append(Arg('indexed_variable', full_text, var_name))
+            arguments.append(Arg('indexed_variable', full_text,
+                                 varname=var_name))
         elif isinstance(argument, (Data_Ref, Proc_Component_Ref)):
-            # A structure dereference e.g. base%arg, base%arg(n) It is
-            # a Proc_Component_Ref if the structure constructor uses
-            # self e.g. self%arg
+            # A structure access e.g. a % b or self % c
```

Comment:
Again, this feels like it is re-inventing what is already in the frontend. Could that be reused instead?

Reply:
Please see earlier comment. The main issue I found is that the individual processing fails if there are missing symbols and I don't see a simple way to sort this out, so I'm suggesting we continue with the current temporary solution.